api.planx.uk/s3/uploadFile.ts

```diff
@@ -33,13 +34,22 @@ export const uploadPrivateFile = async (
   };
 
   await s3.putObject(params).promise();
+  const fileUrl = await buildFileUrl(key, "private")
 
   return {
     file_type: fileType,
-    key,
+    fileUrl,
   };
 };
 
+// Construct an API URL for the uploaded file
+const buildFileUrl = async (key: string, path: "public" | "private") => {
+  const s3 = s3Factory();
+  const s3Url = await new URL(s3.getSignedUrl('getObject', { Key: key }));
```

Comment:
Shouldn't the `await` be placed immediately before the `s3` method?

```diff
- const s3Url = await new URL(s3.getSignedUrl('getObject', { Key: key }));
+ const s3Url = new URL(await s3.getSignedUrl('getObject', { Key: key }));
```

Reply:
Also, won't the URL returned by `s3.getSignedUrl` expire at some point?

Reply:
I believe that this was what I had initially, and I just tried it again.

`const s3Url = new URL(await s3.getSignedUrl('getObject', { Key: key }));` returns a tslint warning `'await' has no effect on the type of this expression.ts(80007)`

Turns out this is because `getSignedUrl()` is actually synchronous - I've not just dropped the `await` altogether.

Docs: https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#getSignedUrl

Reply:
> Also, won't the URL returned by s3.getSignedUrl expire at some point?

Yeah this is an interesting one.

We're only calling `getSignedUrl()` here to get the `URL.pathname` to build a link to our own API, and we use this to get the file key which is later used in `getFileFromS3()`. The fact that the URL will expire doesn't matter - the `pathname` value is consistent.

This feels like a bit of an odd way to just get a key for a file, but it's a reversion to our previous method of doing this and allows us to get the key correctly encoded without worrying about encoding / decoding using the same method that S3 uses.

Also: I'm pretty sure the files are still public on S3 unless I'm mistaken? We're just not exposing that bucket URL any longer, and routing consumers through our API where we check an API key. It's been a while since we originally discussed this but I'd imagine we want to address this right?

Reply:
Yes, we'll want to lock down the S3 bucket so that only our API service has access to it, and all public access will need to go through our API service.


If `pathname` doesn't change then we're good, yes :+1: